Tools: Reoder post-build hooks and bl merging

theotherjimmy opened the pull request:
### Description

On current master, post-build hooks are run before any merging that
managed bootloader mode would do. This puts several targets in a hard
place: they need the post-build hook to run to complute a digest of
the ROM, but when using managed bl mode, these digests are incorrect. 

This PR reorders the post-build hooks to run after the merging of 
managed bootloader mode, and on the "Image" (as opposed to the 
"Update Image"). This should allow post build hooks that generate a
digest of the ROM to compute a correct digest.


### Pull request type

    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@bridadan, For tools changes
@ARMmbed/mbed-os-storage , This team raised the issue

theotherjimmy committed:
Simplify hook tool implementation

The "hook tools" were capable of hooking into many commands run by the build system.
To my knowlage, the only hook is the "post-build-hook". The post build hook could be
easier to reason about if the implementation is specialized for just post-build
hooking.

This commit make it much easier to point out where post build hooks are called by
making the call explicit.

theotherjimmy committed:
Reorder post-build and managed bootloader mode merging

Priously, post-bulid was run before the merge from managed bootloader mode.
This renders many post-build scripts less than useful, as most of them
compute a digest of the ROM image generated in a build.

This reorders the post-build scripts to come after the managed bootloader
mode so that post-build script digests are useful again.

bridadan commented on tools/toolchains/__init__.py:
```diff
@@ -648,18 +679,34 @@ def link_program(self, r, tmp_path, name):
         self.notify.var("binary", filename)
 
-        return full_path, needed_update
+        return full_path, updatable
```
Can we drop `needed_update` completely then? Doesn't seem to be used in this function.

Also, I can't find a call in our code base that actually uses the `updatable` result from `link_program()`. Is it possible to remove it?

theotherjimmy replied:
Yes. We can remove that too. We may have to update the online compiler to use this new API.

bridadan reviewed (approved):
Nicely done!

theotherjimmy committed:
Correct test import locations

A few functions moved into their own module

theotherjimmy committed:
Remove use of hooks from IAR

Aparantely I missed this previously

theotherjimmy committed:
Correct use before definition bug in ARM

theotherjimmy committed:
Drop unused variable from link_program

bridadan reviewed (approved):
Good stuff!

0xc0170 commented on tools/regions.py:
```diff
@@ -0,0 +1,174 @@
+# mbed SDK
+# Copyright (c) 2011-2013 ARM Limited
```
new files should have SPDX identifier and correct year

0xc0170 reviewed (changes requested):
Small change request

yossi2le reviewed (approved):
LGTM

theotherjimmy committed:
Add SPDX identifier

theotherjimmy committed:
Define filename when OUTPUT_EXT is elf

theotherjimmy committed:
Prevent "Image: None" in log

theotherjimmy committed:
Correct all pylint warnings and errors except 4

The 4 remaining are related to the fact that __init__.py defines a base class.
They'll be removed next commit

theotherjimmy committed:
Move toolchain base class into another file

theotherjimmy committed:
Fix formatting erros in arm.py

except for the extremely long regex

theotherjimmy committed:
Fix pylint warnings in gcc.py except for the long regex

theotherjimmy committed:
Fix pylint in iar.py except for the long regex

theotherjimmy committed:
Add SPDX identifier to liscense in "new" file

theotherjimmy committed:
Remove unused import patch
